pkg/identity/gitlabcom/principal.go

```diff
@@ -0,0 +1,205 @@
+	}
+
+	if claims.PipelineID == "" {
+		return nil, errors.New("missing pipeline_id claim in ID token")
+	}
+
+	if claims.JobID == "" {
+		return nil, errors.New("missing job_id claim in ID token")
+	}
+
+	if claims.Ref == "" {
+		return nil, errors.New("missing ref claim in ID token")
+	}
+
+	if claims.RefType == "" {
+		return nil, errors.New("missing ref_type claim in ID token")
+	}
+
+	if claims.NamespacePath == "" {
+		return nil, errors.New("missing namespace_path claim in ID token")
+	}
+
+	if claims.NamespaceID == "" {
+		return nil, errors.New("missing namespace_id claim in ID token")
+	}
+
+	if claims.ProjectID == "" {
+		return nil, errors.New("missing project_id claim in ID token")
+	}
+
+	if claims.Sha == "" {
+		return nil, errors.New("missing sha claim in ID token")
+	}
+
+	if claims.RunnerEnvironment == "" {
+		return nil, errors.New("missing runner_environment claim in ID token")
+	}
+
+	if claims.RunnerID == 0 {
+		return nil, errors.New("missing runner_id claim in ID token")
+	}
+
+	var ref string
+	switch claims.RefType {
+	case "branch":
+		ref = "refs/heads/" + claims.Ref
+	case "tag":
+		ref = "refs/tags/" + claims.Ref
+	default:
+		return nil, fmt.Errorf("unexpected ref_type: %s", claims.RefType)
+	}
+
+	return &jobPrincipal{
+		subject:           token.Subject,
+		issuer:            token.Issuer,
+		url:               `https://gitlab.com/`,
+		eventName:         claims.PipelineSource,
+		pipelineID:        claims.PipelineID,
+		repository:        claims.ProjectPath,
+		ref:               ref,
+		repositoryID:      claims.ProjectID,
+		repositoryOwner:   claims.NamespacePath,
+		repositoryOwnerID: claims.NamespaceID,
+		jobID:             claims.JobID,
+		sha:               claims.Sha,
+		runnerID:          claims.RunnerID,
+		runnerEnvironment: claims.RunnerEnvironment,
+	}, nil
+}
+
+func (p jobPrincipal) Name(_ context.Context) string {
+	return p.subject
+}
+
+func (p jobPrincipal) Embed(_ context.Context, cert *x509.Certificate) error {
+	baseURL, err := url.Parse(p.url)
+	if err != nil {
+		return err
+	}
+
+	// Set workflow ref URL to SubjectAlternativeName on certificate
+	cert.URIs = []*url.URL{baseURL.JoinPath(p.repository, "/-/jobs/", p.jobID)}
+
+	// Embed additional information into custom extensions
+	cert.ExtraExtensions, err = certificate.Extensions{
+		Issuer:                          p.issuer,
+		BuildSignerURI:                  baseURL.JoinPath(p.repository, "/-/jobs/", p.jobID).String(),
+		RunnerEnvironment:               p.runnerEnvironment,
+		SourceRepositoryURI:             baseURL.JoinPath(p.repository).String(),
+		SourceRepositoryDigest:          p.sha,
+		SourceRepositoryRef:             p.ref,
+		SourceRepositoryIdentifier:      p.repositoryID,
+		SourceRepositoryOwnerURI:        baseURL.JoinPath(p.repositoryOwner).String(),
+		SourceRepositoryOwnerIdentifier: p.repositoryOwnerID,
+		BuildConfigURI:                  baseURL.JoinPath(p.repository, "/-/jobs/", p.jobID).String(),
```

Comment:
Missing BuildConfigDigest, which is pipeline sha

Reply:
This isn't implemented yet - https://gitlab.com/gitlab-org/gitlab/-/merge_requests/117923

Reply:
Ah, can we track that on the OID doc then to note it’sa WIP?

Reply:
Sure thing. I'll open another PR.